Can not add entities with ManyToOne relationship to User (with services/DTOs) using Gateway UI

<!--
- Please follow the issue template below for bug reports.
- If you have a support request rather than a bug, please use [Stack Overflow](http://stackoverflow.com/questions/tagged/jhipster) with the JHipster tag.
- For bug reports it is mandatory to run the command `jhipster info` in your project's root folder, and paste the result here.
- Tickets opened without any of these pieces of information will be **closed** without any explanation.
-->

##### **Overview of the issue**

When testing #8560 that fixed #8550, we can not add entities with a ManyToOne relationship to User (with services/DTOs) using the Gateway UI:

```
microservice2-app_1         | 2018-10-15 14:00:29.715 ERROR 1 --- [  XNIO-2 task-6] o.h.i.ExceptionMapperStandardImpl        : HHH000346: Error during managed flush [org.hibernate.TransientPropertyValueException: object references an unsaved transient instance - save the transient instance before flushing : com.example.microservice2.domain.D.user -> com.example.microservice2.domain.User]
microservice2-app_1         | 2018-10-15 14:00:29.720 ERROR 1 --- [  XNIO-2 task-6] o.z.problem.spring.common.AdviceTrait    : Internal Server Error
microservice2-app_1         |
microservice2-app_1         | org.springframework.dao.InvalidDataAccessApiUsageException: org.hibernate.TransientPropertyValueException: object references an unsaved transient instance - save the transient instance before flushing : com.example.microservice2.domain.D.user -> com.example.microservice2.domain.User; nested exception is java.lang.IllegalStateException: org.hibernate.TransientPropertyValueException: object references an unsaved transient instance - save the transient instance before flushing : com.example.microservice2.domain.D.user -> com.example.microservice2.domain.User
microservice2-app_1         | 	at org.springframework.orm.jpa.EntityManagerFactoryUtils.convertJpaAccessExceptionIfPossible(EntityManagerFactoryUtils.java:364)
microservice2-app_1         | 	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.translateExceptionIfPossible(HibernateJpaDialect.java:225)
microservice2-app_1         | 	at org.springframework.orm.jpa.JpaTransactionManager.doCommit(JpaTransactionManager.java:540)
microservice2-app_1         | 	at org.springframework.transaction.support.AbstractPlatformTransactionManager.processCommit(AbstractPlatformTransactionManager.java:746)

.
.
.

microservice2-app_1         | 	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:830)
microservice2-app_1         | 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
microservice2-app_1         | 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
microservice2-app_1         | 	at java.lang.Thread.run(Thread.java:748)
microservice2-app_1         | Caused by: java.lang.IllegalStateException: org.hibernate.TransientPropertyValueException: object references an unsaved transient instance - save the transient instance before flushing : com.example.microservice2.domain.D.user -> com.example.microservice2.domain.User
microservice2-app_1         | 	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:146)
microservice2-app_1         | 	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:157)
microservice2-app_1         | 	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:164)
microservice2-app_1         | 	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1443)
microservice2-app_1         | 	at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:494)
microservice2-app_1         | 	at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:3245)
microservice2-app_1         | 	at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:2451)
microservice2-app_1         | 	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:473)
microservice2-app_1         | 	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.beforeCompletionCallback(JdbcResourceLocalTransactionCoordinatorImpl.java:156)
microservice2-app_1         | 	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.access$100(JdbcResourceLocalTransactionCoordinatorImpl.java:38)
microservice2-app_1         | 	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl$TransactionDriverControlImpl.commit(JdbcResourceLocalTransactionCoordinatorImpl.java:231)
microservice2-app_1         | 	at org.hibernate.engine.transaction.internal.TransactionImpl.commit(TransactionImpl.java:68)
microservice2-app_1         | 	at org.springframework.orm.jpa.JpaTransactionManager.doCommit(JpaTransactionManager.java:536)
microservice2-app_1         | 	... 131 common frames omitted
microservice2-app_1         | Caused by: org.hibernate.TransientPropertyValueException: object references an unsaved transient instance - save the transient instance before flushing : com.example.microservice2.domain.D.user -> com.example.microservice2.domain.User
microservice2-app_1         | 	at org.hibernate.engine.spi.CascadingActions$8.noCascade(CascadingActions.java:380)
microservice2-app_1         | 	at org.hibernate.engine.internal.Cascade.cascade(Cascade.java:142)
microservice2-app_1         | 	at org.hibernate.event.internal.AbstractFlushingEventListener.cascadeOnFlush(AbstractFlushingEventListener.java:150)
microservice2-app_1         | 	at org.hibernate.event.internal.AbstractFlushingEventListener.prepareEntityFlushes(AbstractFlushingEventListener.java:141)
microservice2-app_1         | 	at org.hibernate.event.internal.AbstractFlushingEventListener.flushEverythingToExecutions(AbstractFlushingEventListener.java:74)
microservice2-app_1         | 	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:38)
microservice2-app_1         | 	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1437)
microservice2-app_1         | 	... 140 common frames omitted
microservice2-app_1         |
microservice2-app_1         | 2018-10-15 14:00:29.770  WARN 1 --- [  XNIO-2 task-6] .m.m.a.ExceptionHandlerExceptionResolver : Resolved [org.springframework.dao.InvalidDataAccessApiUsageException: org.hibernate.TransientPropertyValueException: object references an unsaved transient instance - save the transient instance before flushing : com.example.microservice2.domain.D.user -> com.example.microservice2.domain.User; nested exception is java.lang.IllegalStateException: org.hibernate.TransientPropertyValueException: object references an unsaved transient instance - save the transient instance before flushing : com.example.microservice2.domain.D.user -> com.example.microservice2.domain.User]
```

Here is the JDL:

```
application {
  config {
    baseName gateway,
    applicationType gateway,
    authenticationType oauth2,
    packageName com.example.gateway,
    prodDatabaseType postgresql,
    //clientFramework react,
    buildTool gradle
  }
  entities AA, B, C, D
}

application {
  config {
    baseName microservice1,
    applicationType microservice,
    authenticationType oauth2,
    packageName com.example.microservice1,
    prodDatabaseType postgresql,
    buildTool gradle
  }
  entities AA, B
}

application {
  config {
    baseName microservice2,
    applicationType microservice,
    authenticationType oauth2,
    packageName com.example.microservice2,
    prodDatabaseType postgresql,
    buildTool gradle
  }
  entities C, D
}

entity AA {
    data String
}

entity B {
    data String
}

entity C {
    data String
}

entity D {
    data String
}

relationship ManyToOne {
    D{user(login)} to User
}

dto * with mapstruct
service * with serviceClass

microservice AA, B with microservice1
microservice C, D with microservice2
```

![screen shot 2018-10-15 at 10 09 02 am](https://user-images.githubusercontent.com/1197137/46956110-75692e80-d062-11e8-88b4-b2b5688856e0.png)





<!-- Explain the bug, if an error is being thrown a stack trace helps -->

##### **Motivation for or Use Case**

<!-- Explain why this is a bug for you -->

##### **Reproduce the error**

<!-- For bug reports, an unambiguous set of steps to reproduce the error -->

##### **Related issues**

<!-- Has a similar issue been reported before? Please search both closed & open issues -->

##### **Suggest a Fix**

<!-- If you can't fix the bug yourself, perhaps you can point to what might be
  causing the problem (line of code or commit) -->

##### **JHipster Version(s)**

```
gateway@0.0.0 /Users/jsm174/git/jhipster/jh-oauth2-angular-many2one/gateway
└── generator-jhipster@5.4.2

```


##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**


<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "enableHibernateCache": true,
    "enableSwaggerCodegen": false,
    "enableTranslation": true,
    "jhiPrefix": "jhi",
    "languages": [
      "en"
    ],
    "messageBroker": false,
    "nativeLanguage": "en",
    "packageName": "com.example.gateway",
    "packageFolder": "com/example/gateway",
    "prodDatabaseType": "postgresql",
    "searchEngine": false,
    "serviceDiscoveryType": "eureka",
    "skipClient": false,
    "skipServer": false,
    "testFrameworks": [],
    "websocket": false,
    "baseName": "gateway",
    "applicationType": "gateway",
    "authenticationType": "oauth2",
    "buildTool": "gradle",
    "jhipsterVersion": "5.4.2",
    "skipUserManagement": true,
    "clientPackageManager": "npm",
    "cacheProvider": "hazelcast",
    "serverPort": "8080",
    "clientFramework": "angularX",
    "useSass": true
  },
  "entities": [
    "AA",
    "B",
    "C",
    "D"
  ]
}
</pre>
</details>


##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
entity AA {
  data String
}
entity B {
  data String
}
entity C {
  data String
}
entity D {
  data String
}
relationship ManyToOne {
  D{user(login)} to User
}

dto AA, B, C, D with mapstruct
service AA, B, C, D with serviceClass
microservice AA, B with microservice1
microservice C, D with microservice2

</pre>
</details>


##### **Environment and Tools**

java version "1.8.0_181"
Java(TM) SE Runtime Environment (build 1.8.0_181-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.181-b13, mixed mode)

git version 2.17.1 (Apple Git-112)

node: v8.12.0

npm: 6.4.1

yarn: 1.10.1

Docker version 18.06.1-ce, build e68fc7a

docker-compose version 1.22.0, build f46880f

-   [X] Checking this box is mandatory (this is just to show you read everything)

<!-- Love JHipster? Please consider supporting our collective:
👉  https://opencollective.com/generator-jhipster/donate -->

--- a/generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
+++ b/generators/entity-server/templates/src/main/java/package/web/rest/EntityResource.java.ejs
@@ -87,7 +87,7 @@ import org.springframework.web.util.UriComponentsBuilder;
 <%_ if (reactiveRepositories) { _%>
 import org.springframework.http.MediaType;
 <%_ } _%>
-import org.springframework.http.ResponseEntity; <%_ if (!viaService) { %>
+import org.springframework.http.ResponseEntity; <%_ if (!viaService && !saveUserSnapshot) { %>
 import org.springframework.transaction.annotation.Transactional; <% } %>
 import org.springframework.web.bind.annotation.*;
 <%_ if (reactiveRepositories) { _%>
@@ -114,7 +114,7 @@ import static org.elasticsearch.index.query.QueryBuilders.*;<% } %>
  * REST controller for managing {@link <%= packageName %>.domain.<%= asEntity(entityClass) %>}.
  */
 @RestController
-@RequestMapping("/api")<% if (databaseType === 'sql' && !viaService) { %>
+@RequestMapping("/api")<% if (databaseType === 'sql' && !viaService && !saveUserSnapshot) { %>
 @Transactional <%_ } %>
 public class <%= entityClass %>Resource {
 

